pkg/xds/generator/inbound_proxy_generator.go

```diff
@@ -103,3 +116,36 @@ func (g InboundProxyGenerator) Generate(ctx xds_context.Context, proxy *model.Pr
 	}
 	return resources, nil
 }
+
+func (g *InboundProxyGenerator) BuildInboundRoutes(cluster envoy_common.Cluster, rateLimit *mesh_proto.RateLimit) (envoy_common.Routes, error) {
+	route := envoy_common.NewRouteFromCluster(cluster)
+
+	if rateLimit != nil && rateLimit.GetConf().GetHttp() != nil {
+		// Source
+		if len(rateLimit.GetSources()) > 0 {
+			if route.Match == nil {
+				route.Match = &mesh_proto.TrafficRoute_Http_Match{}
+			}
+
+			if route.Match.Headers == nil {
+				route.Match.Headers = make(map[string]*mesh_proto.TrafficRoute_Http_Match_StringMatcher)
+			}
+
+			var selectorRegexs []string
+			for _, selector := range rateLimit.SourceTags() {
+				selectorRegexs = append(selectorRegexs, tags.MatchingRegex(selector))
+			}
+			regexOR := tags.RegexOR(selectorRegexs...)
+
+			route.Match.Headers[v3.TagsHeaderName] = &mesh_proto.TrafficRoute_Http_Match_StringMatcher{
+				MatcherType: &mesh_proto.TrafficRoute_Http_Match_StringMatcher_Regex{
+					Regex: regexOR,
+				},
+			}
+		}
+
+		route.RateLimit = rateLimit
+	}
```

Comment:
Ok, but then we have only one route which has to match the headers that you set.
This means that only service with matching tags will be able to consume it.

Could you add another client to E2E test with different tags and double-check that requests from another client are working and are not rate limitted.